horovod/run/run.py

```diff
@@ -714,28 +789,140 @@ def run():
+        ssh_port=None,
+        disable_cache=None,
+        output_filename=None,
+        verbose=None,
+        use_gloo=None,
+        use_mpi=None):
+    """
+    Launch a Horovod job to run the specified process function and get the return value.
+
+    :param func: The function to be run in Horovod job processes. The function return value will
+                 be collected as the corresponding Horovod process return value.
+    :param args: Arguments to pass to `func`.
+    :param kwargs: Keyword arguments to pass to `func`.
+    :param np: Number of Horovod processes.
+    :param hosts: List of host names and the number of available slots
+                  for running processes on each, of the form: <hostname>:<slots>
+                  (e.g.: host1:2,host2:4,host3:1 indicating 2 processes can run on host1,
+                  4 on host2, and 1 on host3). If not specified, defaults to using localhost:<np>
+    :param hostfile: Path to a host file containing the list of host names and the number of
+                     available slots. Each line of the file must be of the form:
+                     <hostname> slots=<slots>
+    :param start_timeout: Horovodrun has to perform all the checks and
+                          start the processes before the specified
+                          timeout. The default value is 30 seconds.
+                          Alternatively, The environment variable
+                          HOROVOD_START_TIMEOUT can also be used to
+                          specify the initialization timeout.
+    :param ssh_port: SSH port on all the hosts.
+    :param disable_cache: If the flag is not set, horovodrun will perform
+                          the initialization checks only once every 60
+                          minutes -- if the checks successfully pass.
+                          Otherwise, all the checks will run every time
+                          horovodrun is called.'
+    :param output_filename: For Gloo, writes stdout / stderr of all processes to a filename of the form
+                            <output_filename>/rank.<rank>/<stdout | stderr>. The <rank> will be padded with 0
+                            characters to ensure lexicographical order.
+                            For MPI, delegates its behavior to mpirun.
+    :param verbose: If this flag is set, extra messages will be printed.
+    :param use_gloo: Run Horovod using the Gloo controller. This will
+                     be the default if Horovod was not built with MPI support.
+    :param use_mpi: Run Horovod using the MPI controller. This will
+                    be the default if Horovod was built with MPI support.
+
+    :return: Return a list which contains values return by all Horovod processes.
+             The index of the list corresponds to the rank of each Horovod process.
+    """
+
+    def wrapped_func():
+        return func(*args, **kwargs)
+
+    hargs = HorovodArgs()
+
+    hargs.np = np
```

Comment:
simpler if we use `namedtuple` to define `HorovodArgs`.

Reply:
Here there're many args I defined default value None in class `HorovodArgs` constructor. And in `run` we only set few arguments. If use  `namedtuple`, we need to copy the argments list twice completely, like:

```
HorovodArgs = namedtuple('arg1', 'arg2', ..., 'arg100')
hargs = HorovodArgs(arg1=XXX, arg2=XXX, ..., arg100=XXX)

```
but most of the args we only need to keep default value `None`.

Reply:
Unfortunately, in the current design we cannot use `namedtuple` because the downstream `_run` function modifies some of the object's values (therefore, we cannot assume immutability).  We cannot change the behavior to use `namedtuple._replace` as it is currently designed because there command line entrypoint uses the same code path, but passes in parsed arguments object (here @WeichenXu123 is relying on Python's duck typing to get around this interface overloading).

Longterm, I think we should consolidate these two code paths (functional vs command line) to use the same object to represent materialized argument values.  But I don't think it's necessary to do that in this PR since it's an implementation detail not exposed to the end user.